up-l2/rpcclient.adoc

```diff
@@ -22,79 +22,199 @@ SPDX-License-Identifier: Apache-2.0
 
-All link:../languages.adoc[language-specific uProtocol libraries] *MUST* declare the `RpcClient` interface with method:
+=== OnError
+
+An `RpcClient` implementation invokes this method when the invocation of the method was not successful.
 
 [source]
 ----
-invokeMethod(
-  IN method: UUri,
-  IN options: CallOptions,
-  IN payload: UPayload?,
-  OUT result: Future<UMessage, UStatus>
-)
+onError(error: UStatus)
 ----
 
-.Parameters
-[#parameters, width="100%",cols="15%,25%,60%"]
+.onReceive Parameters
+[width="100%",cols="15%,15%,70%"]
 |===
+|Parameter | Type | Description
 
-| Parameter | Type | Description
+| error
+| link:../basics/error_model.adoc[Error Model]
+| Information about the error that has occurred.
 
-| method
-| link:../basics/uri.adoc[UUri]
-| The URI representing the service method to invoke.
+|===
 
-| options
-| CallOptions
-| Additional meta data to include in the request message.
+[mermaid]
+ifdef::env-github[[source,mermaid]]
+----
+sequenceDiagram
 
-| payload
-| link:../basics/upayload.adoc[UPayload]
-| The payload data to include in the request message.
+participant H as handler:ResponseHandler
+participant R as client:RpcClient
 
-| result
-| Future<Void, link:../basics/error_model.adoc[UStatus]>
-| A succeeded future containing the response message returned by the service provider. +
-A failed future indicating an error that occurred while transmitting the request or response message.
-|===
+R-)H : onError(error: UStatus)
+activate H
+deactivate H
+----
 
+== RpcClient
 
-== Implementation
+[#invoke-method]
+=== InvokeMethod
 
-Implementations communicate to their underlining transport through the link:../up-l1/README.adoc#_utransport[uTransport Interface].
+[source]
+----
+invokeMethod(request: UMessage, handler: ResponseHandler)
+----
 
-* Every link:../upclient.adoc[uPClient library] *MUST* implement the `RpcClient` interface.
+.Parameters
+[#parameters, width="100%",cols="15%,25%,60%"]
+|===
+
+| Parameter | Type | Description
+
+| request
+| link:../basics/umessage.adoc[UMessage]
+| The RPC Request message to send to the service provider.
+
+| handler
+| <<response-handler, ResponseHandler>>
+| The handler to invoke with the outcome of the RPC call.
+|===
 
-* The `UUri` and `CallOptions` passed to `invokeMethod()` *MUST* be cached non-persistent memory per call, this is to be able to complete the `Future` when the response is received from the server.
+RpcClient implementations
 
-* The `UPayload` passed to `invokeMethod()` *MAY* be null or empty.
+* *MUST* fail the given handler with a `UCode.DEADLINE_EXCEEDED` if no corresponding Response message has been received before the Request message expires.
+* *MUST* fail the given handler with a `UCode.INVALID_PARAMETER` if the given message is not a link:../basics/uattributes.adoc[valid RPC Request message].
+* *MUST* fail the given handler with a `UCode.ALREADY_EXISTS` if a service invocation with the same request message is currently pending and not expired.
```

Comment:
I really like this point!

> I really think this section should be revised to list the kinds of error states that should be checked for when called (i.e. invalid parameter or pre-existing request ID) and conditions that need to be monitored for before the response arrives (i.e. deadline exceeded).

Reply:
Over on #151, I provided an example of one way to describe errors and conditions as behaviors rather than explicit error codes: https://github.com/eclipse-uprotocol/up-spec/pull/151#issuecomment-2123565274